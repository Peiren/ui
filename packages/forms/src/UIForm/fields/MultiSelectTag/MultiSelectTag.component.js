import PropTypes from 'prop-types';
import React from 'react';
import keycode from 'keycode';
import Typeahead from 'react-talend-components/lib/Typeahead';
import FieldTemplate from '../FieldTemplate';
import Tags from './Tags.component';

import theme from './MultiSelectTagWidget.scss';

const INPUT_TEXT_INDENT = 7.5;
const TAGS_HEIGHT = 28;
const TAGS_MARGIN_TOP = 5;
const INPUT_MIN_WIDTH = 135;

function escapeRegexCharacters(str) {
	return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function getNewItemText(value) {
	return `${value} (new)`;
}

export default class MultiSelectTag extends React.Component {
	constructor(props) {
		super(props);
		this.state = { value: '' };
		this.theme = {
			container: theme.typeahead,
			itemsContainer: theme['items-container'],
			itemsList: theme.items,
		};

		this.onChange = this.onChange.bind(this);
		this.onFocus = this.onFocus.bind(this);
		this.onInputMount = this.onInputMount.bind(this);
		this.onKeyDown = this.onKeyDown.bind(this);
		this.onRemoveTag = this.onRemoveTag.bind(this);
		this.onAddTag = this.onAddTag.bind(this);
		this.onTagsMount = this.onTagsMount.bind(this);
		this.resetSuggestions = this.resetSuggestions.bind(this);
	}

	/**
	 * Update input spacing on mount
	 */
	componentDidMount() {
		this.updateSpaces();
	}

	/**
	 * On Tags value change
	 * - Update input spacing
	 * - Update suggestions if they are displayed
	 * @param value The tags values
	 */
	componentDidUpdate({ value }) {
		if (!this.input || !this.tags || value === this.props.value) {
			return;
		}
		this.updateSpaces();
		if (this.state.suggestions) {
			this.updateSuggestions();
		}
	}

	/**
	 * Save input ref
	 * @param input The input ref
	 */
	onInputMount(input) {
		this.input = input;
	}

	/**
	 * Save tags list ref
	 * @param tags The tags list ref
	 */
	onTagsMount(tags) {
		this.tags = tags;
	}

	/**
	 * Manage suggestion selection
	 * @param event
	 * @param focusedItemIndex
	 * @param newFocusedItemIndex
	 */
	onKeyDown(event, { focusedItemIndex, newFocusedItemIndex }) {
		switch (event.which) {
		case keycode.codes.enter:
			if (!this.state.suggestions) {
				break;
			}
			event.preventDefault();
			// suggestions are displayed and an item has the focus : we select it
			if (Number.isInteger(focusedItemIndex)) {
				this.onAddTag(event, { itemIndex: focusedItemIndex });
			}
			break;
		case keycode.codes.down:
		case keycode.codes.up:
			event.preventDefault();
			this.setState({ focusedItemIndex: newFocusedItemIndex });
			break;
		default:
			break;
		}
	}

	/**
	 * Update suggestions on input value change
	 * @param event The input change event
	 * @param value The new input value
	 */
	onChange(event, { value }) {
		this.updateSuggestions(value);
	}

	/**
	 * Update suggestions on input focus
	 */
	onFocus() {
		this.updateSuggestions();
	}

	/**
	 * Add a new tag
	 * @param event The user event
	 * @param itemIndex The selected suggestion index
	 */
	onAddTag(event, { itemIndex }) {
		const currentValue = this.state.value;
		const selectedOption = this.state.suggestions[itemIndex];
		const isCreation = getNewItemText(currentValue) === selectedOption;

		const newValue = isCreation ? currentValue : selectedOption;
		const payload = {
			schema: this.props.schema,
			value: this.props.value.concat(newValue),
		};
		this.props.onChange(event, payload);
		this.props.onFinish(event, payload);
		this.updateSuggestions('');
	}

	/**
	 * Remove a tag
	 * @param event The user event
	 * @param itemIndex The tag index
	 */
	onRemoveTag(event, itemIndex) {
		const value = this.props.value.slice(0);
		value.splice(itemIndex, 1);
		const payload = { schema: this.props.schema, value };
		this.props.onChange(event, payload);
		this.props.onFinish(event, payload);
	}

	/**
	 * Remove all suggestions
	 */
	resetSuggestions() {
		this.setState({
			suggestions: undefined,
			focusedItemIndex: undefined,
		});
	}

	/**
	 * Add spaces on input to place it after the tags
	 */
	updateSpaces() {
		if (this.props.value.length) {
			const lastTag = this.tags.querySelector('div.tc-badge:last-child');
			const paddingLeft = lastTag.offsetLeft + lastTag.offsetWidth + INPUT_TEXT_INDENT;
			let paddingTop = lastTag.offsetTop - TAGS_MARGIN_TOP;
			const toNextLine = (this.input.offsetWidth - paddingLeft) < INPUT_MIN_WIDTH;
			if (toNextLine) {
				paddingTop += TAGS_HEIGHT;
				this.input.style.paddingLeft = '0px';
			} else {
				this.input.style.paddingLeft = `${paddingLeft}px`;
			}
			this.input.style.marginTop = `${paddingTop}px`;
		} else {
			this.input.style.paddingLeft = '0px';
			this.input.style.paddingTop = '0px';
		}
	}

	/**
	 * Update suggestions
	 * - remove current tags values
	 * - filter based on input value
	 * @param value The new input value. Undefined if we should use the current input value.
	 */
	updateSuggestions(value) {
		const currentValue = value === undefined ? this.state.value : value;
		let suggestions = this.props.schema.titleMap
			.map(item => item.value)
			.filter(item => this.props.value.indexOf(item) < 0);

		if (currentValue) {
			const escapedValue = escapeRegexCharacters(currentValue.trim());
			const regex = new RegExp(escapedValue, 'i');
			suggestions = suggestions.filter(itemValue => regex.test(itemValue));

			if (!suggestions.length && this.props.schema.restricted === false) {
				suggestions.push(getNewItemText(currentValue));
			}
		}

		this.setState({
			focusedItemIndex: suggestions.length ? 0 : undefined,
			suggestions,
			value: currentValue,
		});
	}

	render() {
		const { id, isValid, errorMessage, schema } = this.props;

		return (
			<FieldTemplate
				description={schema.description}
				errorMessage={errorMessage}
				id={id}
				isValid={isValid}
				label={schema.title}
			>
				<div className={theme.wrapper}>
					<Typeahead
						id={id}
						autoFocus={schema.autoFocus || false}
						disabled={schema.disabled || false}
						focusedItemIndex={this.state.focusedItemIndex}
						items={this.state.suggestions}
						multiSection={false}
						onBlur={this.resetSuggestions}
						onChange={this.onChange}
						onFocus={this.onFocus}
						inputRef={this.onInputMount}
						onKeyDown={this.onKeyDown}
						onSelect={this.onAddTag}
						placeholder={schema.placeholder}
						readOnly={schema.readOnly || false}
						theme={this.theme}
						value={this.state.value}
					/>
					<div className={theme.caret}>
						<span className="caret" />
					</div>
					<Tags
						onRemoveTag={this.onRemoveTag}
						onTagsMount={this.onTagsMount}
						readonly={schema.readOnly || schema.disabled}
						titleMap={schema.titleMap}
						value={this.props.value}
					/>
				</div>
			</FieldTemplate>
		);
	}
}

if (process.env.NODE_ENV !== 'production') {
	MultiSelectTag.propTypes = {
		id: PropTypes.string,
		isValid: PropTypes.bool,
		errorMessage: PropTypes.string,
		onChange: PropTypes.func.isRequired,
		onFinish: PropTypes.func.isRequired,
		schema: PropTypes.shape({
			autoFocus: PropTypes.bool,
			description: PropTypes.string,
			disabled: PropTypes.bool,
			placeholder: PropTypes.string,
			readOnly: PropTypes.bool,
			restricted: PropTypes.bool,
			title: PropTypes.string,
			titleMap: PropTypes.arrayOf(PropTypes.shape({
				name: PropTypes.string.isRequired,
				value: PropTypes.string.isRequired,
			})),
		}),
		value: PropTypes.arrayOf(PropTypes.string),
	};
}

MultiSelectTag.defaultProps = {
	isValid: true,
	schema: {},
	value: [],
};
