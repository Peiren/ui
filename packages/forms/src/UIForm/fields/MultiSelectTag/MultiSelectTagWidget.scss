$tf-datalist-items-background-color: $white !default;
$tf-datalist-items-shadow-color: $shadow !default;
$tf-typeahead-input-width: 135px !default;


.wrapper {
	position: relative;
	display: flex;
	flex-direction: row-reverse;

	.typeahead {
		width: 100%;

		> div {
			width: 100%;
		}

		.items-container {
			position: absolute;
			margin-top: $padding-smaller;
			top: 100%;
			right: 0;
			left: 0;
			max-height: 70vh;
			min-width: $tf-typeahead-input-width;
			font-size: 1.5rem;
			background-color: $tf-datalist-items-background-color;
			border-radius: $border-radius-base;
			box-shadow: 0 1px 3px 0 $tf-datalist-items-shadow-color;
			overflow-y: auto;
			z-index: 999;
		}

		.items {
			margin: 0;
			padding: 0;
			list-style-type: none;

			> li > div {
				padding: $padding-smaller;
			}
		}
	}

	.caret {
		position: absolute;
		bottom: $padding-smaller;
		right: $padding-smaller;
	}

	.tags {
		position: absolute;
		top: 0;
		left: 0;
		height: 28px;
		pointer-events: none;

		:global(.tc-badge) {
			pointer-events: auto;
		}
	}
}
