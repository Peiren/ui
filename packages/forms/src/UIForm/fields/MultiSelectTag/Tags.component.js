import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import Badge from 'react-talend-components/lib/Badge';

import theme from './MultiSelectTagWidget.scss';

function getLabel(titleMap, value) {
	const itemConf = titleMap.find(item => item.value === value);
	return itemConf ? itemConf.name : value;
}

export default function Tags({ onRemoveTag, onTagsMount, readonly, titleMap, value }) {
	return (
		<div
			className={classNames(theme.tags, 'tags')}
			ref={onTagsMount}
		>
			{
				value.map((val, index) => {
					const label = getLabel(titleMap, val);
					const badgeProps = { label, key: index };
					if (!readonly) {
						badgeProps.onDelete = event => onRemoveTag(event, index);
					}
					return (
						<Badge {...badgeProps} />
					);
				})
			}
		</div>
	);
}
if (process.env.NODE_ENV !== 'production') {
	Tags.propTypes = {
		onRemoveTag: PropTypes.func,
		onTagsMount: PropTypes.func,
		readonly: PropTypes.bool,
		titleMap: PropTypes.arrayOf(PropTypes.shape({
			name: PropTypes.string,
			value: PropTypes.string,
		})).isRequired,
		value: PropTypes.arrayOf(PropTypes.string).isRequired,
	};
}
