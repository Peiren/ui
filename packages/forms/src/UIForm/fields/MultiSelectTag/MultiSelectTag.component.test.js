import React from 'react';
import { shallow } from 'enzyme';
import MultiSelectTag from './MultiSelectTag.component';

describe('MultiSelectTag field', () => {
	const props = {
		id: 'my-select-tag',
		isValid: true,
		errorMessage: 'This is wrong',
		onChange: jest.fn(),
		onFinish: jest.fn(),
		schema: {
			autoFocus: true,
			description: 'This is the MultiSelectTag field',
			disabled: false,
			placeholder: 'Type here',
			readOnly: false,
			restricted: false,
			title: 'Tags',
			titleMap: [{ name: 'toto', value: 'titi' }, { name: 'tata', value: 'tutu' }],
		},
		value: ['aze', 'tutu'],
	};

	it('should render MultiSelectTag', () => {
		// when
		const wrapper = shallow(<MultiSelectTag {...props} />);

		// then
		expect(wrapper.node).toMatchSnapshot();
	});
});
