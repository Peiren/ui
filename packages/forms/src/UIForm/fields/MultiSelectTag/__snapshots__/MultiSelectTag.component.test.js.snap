// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectTag field should render MultiSelectTag 1`] = `
<FieldTemplate
  description="This is the MultiSelectTag field"
  errorMessage="This is wrong"
  id="my-select-tag"
  isValid={true}
  label="Tags"
>
  <div
    className={undefined}
  >
    <Typeahead
      autoFocus={true}
      disabled={false}
      focusedItemIndex={undefined}
      id="my-select-tag"
      inputRef={[Function]}
      items={null}
      multiSection={false}
      noResultText="No result."
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onSelect={[Function]}
      placeholder="Type here"
      position="left"
      readOnly={false}
      searching={false}
      searchingText="Searching for matchesâ€¦"
      theme={
        Object {
          "container": undefined,
          "itemsContainer": undefined,
          "itemsList": undefined,
        }
      }
      value=""
    />
    <div
      className={undefined}
    >
      <span
        className="caret"
      />
    </div>
    <Tags
      onRemoveTag={[Function]}
      onTagsMount={[Function]}
      readonly={false}
      titleMap={
        Array [
          Object {
            "name": "toto",
            "value": "titi",
          },
          Object {
            "name": "tata",
            "value": "tutu",
          },
        ]
      }
      value={
        Array [
          "aze",
          "tutu",
        ]
      }
    />
  </div>
</FieldTemplate>
`;
